# Change directory

In [1]:
pwd

'/Users/chengdan'

In [3]:
cd /Users/chengdan/Desktop/Python_Project_Dan_Cheng

/Users/chengdan/Desktop/Python_Project_Dan_Cheng


In [4]:
pwd

'/Users/chengdan/Desktop/Python_Project_Dan_Cheng'

# Process original data

In [5]:
import pandas as pd

Read "CCDS.20160908_origin.txt" as csv file:

In [7]:
data0 = pd.read_csv("CCDS.20160908_origin.txt",sep="\t")http://localhost:8889/notebooks/Desktop/Python_Project_Dan_Cheng.ipynb#Check-the-type-of-data0

Check the type of data0:

In [8]:
type(data0)

pandas.core.frame.DataFrame

Check the shape of data0:

In [9]:
data0.shape

(34297, 11)

This data frame contains 34297 rows and 11 columns

Check the firs 5 rows of data0:

In [10]:
data0.head()

#chromosome  nc_accession       gene  gene_id  ccds_id ccds_status  \
0           1   NC_000001.8  LINC00115    79854  CCDS1.1   Withdrawn   
1           1  NC_000001.11     SAMD11   148398  CCDS2.2      Public   
2           1  NC_000001.11      NOC2L    26155  CCDS3.1      Public   
3           1  NC_000001.11    PLEKHN1    84069  CCDS4.1      Public   
4           1  NC_000001.11       HES4    57801  CCDS5.1      Public   

  cds_strand cds_from  cds_to  \
0          -        -       -   
1          +   925941  944152   
2          -   944693  959239   
3          +   966531  974574   
4          -   999058  999972   

                                       cds_locations match_type  
0                                                  -       None  
1  [925941-926012, 930154-930335, 931038-931088, ...  Identical  
2  [944693-944799, 945056-945145, 945517-945652, ...  Identical  
3  [966531-966613, 966703-966802, 970276-970422, ...  Identical  
4  [999058-999431, 999525-999612, 999691-999786, ...  Identical

Now I can see that the column labels including: "#chromosome", "nc_accession", "gene", "gene_id", "ccds_id",
"ccds_status", "cds_strand", "cds_from", "cds_to", "cds_locations", "match_type"

Check the composition of the "match_type" column of data0:

In [11]:
data0.iloc[:,-1].value_counts()

Identical    34031
None           219
Partial         47
Name: match_type, dtype: int64

I only want to take the data whose "match_type" is "Identical"

Check the composition of the "ccds_status" column of data0:

In [12]:
data0.iloc[:,5].value_counts()

Public                                32534
Withdrawn                              1731
Under review, withdrawal                 15
Reviewed, update pending                  7
Withdrawn, inconsistent annotation        6
Under review, update                      4
Name: ccds_status, dtype: int64

I only want to take the data whose "ccds_status" is "Public"

In [13]:
import numpy as np

Take the data with "match_type"=="Identical" and "ccds_status"=="Public", and save as data1:

In [26]:
data1 = np.logical_and(data0.iloc[:,5]=="Public",data0.iloc[:,-1]=="Identical")

Check the type of data1:

In [18]:
type(data1)

pandas.core.series.Series

Now the type of the data is changed from dataframe to series

Check how many genes meet the requirements "match_type"=="Identical" and "ccds_status"=="Public":

In [27]:
data1.value_counts()

True     32533
False     1764
Name: ccds_status, dtype: int64

There are 32533 genes meet the requirements "match_type"=="Identical" and "ccds_status"=="Public"

In [33]:
data2 = data0[data1]

Check data2 "match_type":

In [34]:
data2.match_type.value_counts()

Identical    32533
Name: match_type, dtype: int64

Check data2 "ccds_status":

In [45]:
data2.ccds_status.value_counts()

Public    32533
Name: ccds_status, dtype: int64

Drop the "match_type" column and "ccds_status" column:

In [42]:
data3=data2.drop(["match_type", "ccds_status"],axis=1)

Check the shape of data3:

In [43]:
data3.shape

(32533, 9)

In [52]:
def sum_exon(tmp):
    tmp1=tmp.strip("[").strip("]").split(", ")
    Lcds=sum([np.diff(list(map(int,tmp1[i].split("-"))))[0]+1 for i in range(len(tmp1))])
    return(Lcds)

In [55]:
tmp=data3.iloc[0,-1]

In [56]:
sum_exon(tmp)

2046

In [57]:
cds_len = list(map(sum_exon,data3.cds_locations))

In [58]:
print(cds_len)

[2046, 2250, 1836, 666, 498, 597, 1215, 768, 726, 834, 1047, 990, 780, 828, 624, 2505, 912, 1803, 2013, 1329, 600, 450, 1338, 195, 1947, 1905, 585, 801, 1023, 630, 1359, 1779, 543, 2187, 1494, 1041, 2346, 687, 1134, 2130, 819, 1383, 1911, 1971, 1752, 1017, 732, 1236, 1104, 1062, 5865, 387, 1905, 855, 1830, 1143, 1053, 990, 2565, 1254, 1281, 1143, 1119, 705, 3021, 2109, 2526, 1764, 2067, 903, 528, 1680, 303, 3606, 420, 375, 768, 570, 1389, 4701, 1305, 1539, 1506, 1008, 2376, 822, 966, 3135, 2916, 246, 573, 840, 405, 3522, 5313, 3462, 1452, 495, 417, 318, 996, 807, 3501, 1245, 2061, 558, 909, 2583, 7650, 1041, 1017, 891, 675, 768, 882, 636, 1050, 480, 1971, 2610, 456, 405, 1905, 2184, 1167, 1788, 1386, 909, 585, 1425, 1425, 5157, 5049, 2328, 762, 723, 807, 810, 1251, 1002, 1059, 1854, 915, 1122, 10995, 2412, 510, 2064, 2064, 2931, 2409, 777, 1410, 792, 552, 3543, 843, 1998, 1992, 1995, 1992, 1569, 3840, 1101, 1404, 2334, 1563, 2520, 1692, 15552, 2982, 720, 996, 1080, 876, 546, 1323, 1905

In [59]:
data3.columns.values.tolist()

['#chromosome',
 'nc_accession',
 'gene',
 'gene_id',
 'ccds_id',
 'cds_strand',
 'cds_from',
 'cds_to',
 'cds_locations']

In [60]:
data3["cds_len"] = cds_len

In [61]:
data3.shape

(32533, 10)

In [62]:
data3.columns.values.tolist()

['#chromosome',
 'nc_accession',
 'gene',
 'gene_id',
 'ccds_id',
 'cds_strand',
 'cds_from',
 'cds_to',
 'cds_locations',
 'cds_len']

In [64]:
data3.head()

#chromosome  nc_accession     gene  gene_id  ccds_id cds_strand cds_from  \
1           1  NC_000001.11   SAMD11   148398  CCDS2.2          +   925941   
2           1  NC_000001.11    NOC2L    26155  CCDS3.1          -   944693   
3           1  NC_000001.11  PLEKHN1    84069  CCDS4.1          +   966531   
4           1  NC_000001.11     HES4    57801  CCDS5.1          -   999058   
5           1  NC_000001.11    ISG15     9636  CCDS6.1          +  1013573   

    cds_to                                      cds_locations  cds_len  
1   944152  [925941-926012, 930154-930335, 931038-931088, ...     2046  
2   959239  [944693-944799, 945056-945145, 945517-945652, ...     2250  
3   974574  [966531-966613, 966703-966802, 970276-970422, ...     1836  
4   999972  [999058-999431, 999525-999612, 999691-999786, ...      666  
5  1014477                 [1013573-1013575, 1013983-1014477]      498

In [65]:
data3.sort_values(by=["cds_len"],ascending=False,inplace=True)

In [67]:
data3.head(10)

#chromosome  nc_accession   gene  gene_id      ccds_id cds_strand  \
29050           2  NC_000002.12    TTN     7273  CCDS59435.1          -   
31519           2  NC_000002.12    TTN     7273  CCDS74610.1          -   
25664           2  NC_000002.12    TTN     7273  CCDS54424.1          -   
25662           2  NC_000002.12    TTN     7273  CCDS54422.1          -   
25663           2  NC_000002.12    TTN     7273  CCDS54423.1          -   
25661           2  NC_000002.12    TTN     7273  CCDS54421.1          -   
25470          19  NC_000019.10  MUC16    94025  CCDS54212.1          -   
28842           1  NC_000001.11  OBSCN    84033  CCDS59204.1          +   
5206            6  NC_000006.12  SYNE1    23345   CCDS5236.2          -   
5205            6  NC_000006.12  SYNE1    23345   CCDS5235.1          -   

        cds_from     cds_to  \
29050  178527011  178804641   
31519  178527011  178804641   
25664  178527011  178804641   
25662  178527011  178804641   
25663  178527011  178804641   
25661  178527011  178804641   
25470    8848931    8981137   
28842  228211783  228378794   
5206   152122435  152628330   
5205   152122435  152628330   

                                           cds_locations  cds_len  
29050  [178527011-178527306, 178527445-178527747, 178...   107976  
31519  [178527011-178527306, 178527445-178527747, 178...   103053  
25664  [178527011-178527306, 178527445-178527747, 178...   100272  
25662  [178527011-178527306, 178527445-178527747, 178...    81357  
25663  [178527011-178527306, 178527445-178527747, 178...    81156  
25661  [178527011-178527306, 178527445-178527747, 178...    80781  
25470  [8848931-8849029, 8851275-8851354, 8851677-885...    43524  
28842  [228211783-228212770, 228213440-228213709, 228...    26772  
5206   [152122435-152122675, 152130719-152130777, 152...    26394  
5205   [152122435-152122675, 152130719-152130777, 152...    26250

In [68]:
data3["dup"] = data3.duplicated(subset=["gene"],keep="first")

In [69]:
data3.head()

#chromosome  nc_accession gene  gene_id      ccds_id cds_strand  \
29050           2  NC_000002.12  TTN     7273  CCDS59435.1          -   
31519           2  NC_000002.12  TTN     7273  CCDS74610.1          -   
25664           2  NC_000002.12  TTN     7273  CCDS54424.1          -   
25662           2  NC_000002.12  TTN     7273  CCDS54422.1          -   
25663           2  NC_000002.12  TTN     7273  CCDS54423.1          -   

        cds_from     cds_to  \
29050  178527011  178804641   
31519  178527011  178804641   
25664  178527011  178804641   
25662  178527011  178804641   
25663  178527011  178804641   

                                           cds_locations  cds_len    dup  
29050  [178527011-178527306, 178527445-178527747, 178...   107976  False  
31519  [178527011-178527306, 178527445-178527747, 178...   103053   True  
25664  [178527011-178527306, 178527445-178527747, 178...   100272   True  
25662  [178527011-178527306, 178527445-178527747, 178...    81357   True  
25663  [178527011-178527306, 178527445-178527747, 178...    81156   True

if the gene is occur first time, return False, if the gene is occur more than once,return True

In [81]:
data4 = data3[data3.dup==False]

In [82]:
data4.head()

#chromosome  nc_accession   gene  gene_id      ccds_id cds_strand  \
29050           2  NC_000002.12    TTN     7273  CCDS59435.1          -   
25470          19  NC_000019.10  MUC16    94025  CCDS54212.1          -   
28842           1  NC_000001.11  OBSCN    84033  CCDS59204.1          +   
5206            6  NC_000006.12  SYNE1    23345   CCDS5236.2          -   
31511           2  NC_000002.12    NEB     4703  CCDS74588.1          -   

        cds_from     cds_to  \
29050  178527011  178804641   
25470    8848931    8981137   
28842  228211783  228378794   
5206   152122435  152628330   
31511  151485759  151733155   

                                           cds_locations  cds_len    dup  
29050  [178527011-178527306, 178527445-178527747, 178...   107976  False  
25470  [8848931-8849029, 8851275-8851354, 8851677-885...    43524  False  
28842  [228211783-228212770, 228213440-228213709, 228...    26772  False  
5206   [152122435-152122675, 152130719-152130777, 152...    26394  False  
31511  [151485759-151485932, 151489970-151490076, 151...    25683  False

In [80]:
data4.shape

(18882, 11)

In [83]:
datafinal=data4.drop("dup",axis=1)

In [84]:
datafinal.head()

#chromosome  nc_accession   gene  gene_id      ccds_id cds_strand  \
29050           2  NC_000002.12    TTN     7273  CCDS59435.1          -   
25470          19  NC_000019.10  MUC16    94025  CCDS54212.1          -   
28842           1  NC_000001.11  OBSCN    84033  CCDS59204.1          +   
5206            6  NC_000006.12  SYNE1    23345   CCDS5236.2          -   
31511           2  NC_000002.12    NEB     4703  CCDS74588.1          -   

        cds_from     cds_to  \
29050  178527011  178804641   
25470    8848931    8981137   
28842  228211783  228378794   
5206   152122435  152628330   
31511  151485759  151733155   

                                           cds_locations  cds_len  
29050  [178527011-178527306, 178527445-178527747, 178...   107976  
25470  [8848931-8849029, 8851275-8851354, 8851677-885...    43524  
28842  [228211783-228212770, 228213440-228213709, 228...    26772  
5206   [152122435-152122675, 152130719-152130777, 152...    26394  
31511  [151485759-151485932, 151489970-151490076, 151...    25683

In [85]:
datafinal.shape

(18882, 10)

In [88]:
total_length_CDS = sum(datafinal.cds_len)

In [87]:
print("Number of protein-coding genes counted: ",datafinal.shape[0])

Number of protein-coding genes counted:  18882


In [89]:
print("Total length of human CDS (of longest transcripts) is: ",total_length_CDS, "bp")

Total length of human CDS (of longest transcripts) is:  32899880 bp


In [ ]:
Test:

In [90]:
import numpy as np
import pandas as pd

def sum_exon(cds_locations):
    cds_locations = cds_locations.strip("[").strip("]").split(", ")
    Lcds = sum([np.diff(list(map(int,cds_locations[i].split("-"))))[0]+1 for i in range(len(cds_locations))])
    return(Lcds)

def test_sum_exon():
    
    cds0 = pd.read_csv("./CCDS.20160908.txt_origin",sep="\t") #34297
    
    cds = cds0[np.logical_and(cds0.match_type=="Identical", cds0.ccds_status=="Public")].drop(['match_type','ccds_status'],axis=1)
    
    cds_len = list(map(sum_exon,cds.cds_locations))
    
    cds["cds_len"] = cds_len
    
    cds.columns.values.tolist()
    
    cds.sort_values(by=["cds_len"],ascending=False,inplace=True)
    
    cds["dup"] = cds.duplicated(subset=["gene"],keep="first")
    
    cds = cds[cds.dup==False].drop(labels="dup",axis=1)
    
    total_len_cds = sum(cds.cds_len)
    
    print("Number of protein-coding genes counted: ",cds.shape[0])
    print("Total length of human CDS (of longest transcripts) is: ",total_len_cds, "bp")
    
    assert total_len_cds == 32899880
